### PR TITLE
Approximate non-ASCII characters

### PR DESCRIPTION
**Pull request by andrewgarner**
The font we use to render output documents has a limited character set so we need to remove unsupported characters.

### CODE REVIEW

File: spec/domain/output_document_spec.rb [hunk 1]
```diff
@@ -180,10 +207,17 @@
     subject { output_document.lead }
 
     it do
-      is_expected.to eq(
-        'You recently had a Pension Wise guidance appointment with James ' \
-        "from The Pensions Advisory Service on #{appointment_date}."
-      )
+      is_expected.to eq('You recently had a Pension Wise guidance appointment with James ' \
+                        "from The Pensions Advisory Service on #{appointment_date}.")
```
Comment (stevenwilkin) [inline spec/domain/output_document_spec.rb]: This re-formating should strictly be included in this PR

Comment (stevenwilkin) [inline spec/domain/output_document_spec.rb]: *shouldn't

Comment (andrewgarner) [inline spec/domain/output_document_spec.rb]: OK. I'll follow the existing formatting in the new example introduced.

**Comment by stevenwilkin** [comment]
Looks good, as long as we're ok with adding the transliteration into the domain object and not creating a new one to add in this new functionality

**Comment by gjvis** [comment]
Can we pass in the transliterator, and test this?

**Comment by gjvis** [comment]
And as @stevenwilkin suggests, perhaps the transliteration should happen in another object that wraps an `OutputDocument`?

**Comment by andrewgarner** [comment]
The transliterator comes from `I18n` - it's already well tested.

**Comment by stevenwilkin** [comment]
@gjvis is referring to the object that *applies* the transliteration, not the object *implementing* the transliteration @andrewgarner

**Comment by gjvis** [comment]
I was, but I somehow missed that there were already specs on this PR. Need more coffee... :coffee: :coffee: :coffee:

**Comment by gjvis** [comment]
I'm worried that `OutputDocument` is becoming a bit of a god-object, this does seem like the simplest thing to do to get this functionality in though, so might be too early to extract transliteration into its own thing. Happy to go with the consensus.

**Comment by andrewgarner** [comment]
I've broken the concern of transliteration into a dedicated object and updated the acceptance tests to include transliteration.

cc/ @stevenwilkin @gjvis

File: features/step_definitions/covering_letter_steps.rb [hunk 2]
```diff
@@ -3,17 +3,5 @@
...
-    attendee_address_line_3: output_document.attendee_address_line_3,
-    attendee_town: output_document.attendee_town,
-    attendee_county: output_document.attendee_county,
-    attendee_postcode: output_document.attendee_postcode,
-    attendee_name: output_document.attendee_name,
-    lead: output_document.lead
-  }
-
-  expect_uploaded_csv_to_include(expected)
+  # behaviour is outside of the system, nothing to assert
```
Comment (gjvis) [inline features/step_definitions/covering_letter_steps.rb]: I misunderstood this when we were discussing in person – the original assertion is still valid, we need to make sure the CSV contains all details needed to render a covering letter.

File: features/step_definitions/record_of_guidance_contents_steps.rb [hunk 3]
```diff
@@ -120,12 +120,7 @@
 end
 
 Then(/^the record of guidance should include their details$/) do
-  output_document = OutputDocument.new(@appointment_summary)
-
-  expected = {
-    attendee_name: output_document.attendee_name,
-    value_of_pension_pots: output_document.value_of_pension_pots
-  }
+  expected = fixture(:populated_csv)
```
Comment (gjvis) [inline features/step_definitions/record_of_guidance_contents_steps.rb]: we should assert only the fields we care about, not the entire csv row, otherwise unrelated changes will fail this assertion

File: features/step_definitions/record_of_guidance_contents_steps.rb [hunk 4]
```diff
@@ -135,16 +130,12 @@
...
 
 Then(/^the record of guidance should include the details of the appointment$/) do
-  output_document = OutputDocument.new(@appointment_summary)
-
-  expected = {
-    appointment_date: output_document.appointment_date,
-    guider_first_name: output_document.guider_first_name,
-    guider_organisation: output_document.guider_organisation
-  }
+  expected = fixture(:populated_csv)
```
Comment (gjvis) [inline features/step_definitions/record_of_guidance_contents_steps.rb]: we should assert only the fields we care about, not the entire csv row, otherwise unrelated changes will fail this assertion

File: spec/models/transliterated_appointment_summary_spec.rb [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  describe '.guider_name' do
+    subject { transliterated_appointment_summary.guider_name }
+
+    let(:guider_name) { 'García' }
+
+    it 'approximates non-ASCII characters' do
+      is_expected.to eq('Garcia')
+    end
+  end
```
Comment (gjvis) [inline spec/models/transliterated_appointment_summary_spec.rb]: Some sort of shared example might be useful?

```ruby
shared_examples 'transliterates' do |field|
  subject do
    summary = AppointmentSummary.new(field => 'àéïöü')
    summary.public_send(field)
  end

  it 'approximates non-ASCII characters' do
    is_expected.to eq('aeiou')
  end
end


```

**Comment by gjvis** [comment]
Structure is :+1:, much cleaner. Added a few small comments about tests

File: spec/models/transliterated_appointment_summary_spec.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  describe '.first_name' do
+    subject { transliterated_appointment_summary.first_name }
+
+    let(:first_name) { 'Noemí' }
+
+    it 'approximates non-ASCII characters' do
+      is_expected.to eq('Noemi')
+    end
```
Comment (stevenwilkin) [inline spec/models/transliterated_appointment_summary_spec.rb]: Better to use `it { is_expected.to eq(..) }` for all these assertions?

Comment (gjvis) [inline spec/models/transliterated_appointment_summary_spec.rb]: My 2 cents: the example doesn't really communicate the intention, so I quite like the clarity of the message (in this case)

**Comment by jiannis** [comment]
_Minor point_: might be worth testing out letters such as the [German ß](http://en.wikipedia.org/wiki/ß) (option + s) which doesn't have a direct one letter ASCII equivalent but should be transliterated as “ss”.

File: spec/models/transliterated_appointment_summary_spec.rb (hunk 5 above)
Comment (gjvis) [inline spec/models/transliterated_appointment_summary_spec.rb]: (danger: that code was GitHub-coded :skull: :warning:)

File: spec/models/transliterated_appointment_summary_spec.rb (hunk 6 above)
Comment (stevenwilkin) [inline spec/models/transliterated_appointment_summary_spec.rb]: My complaint was seeing `approximates non-ASCII characters` repeated down the file

File: features/step_definitions/covering_letter_steps.rb (hunk 2 above)
Comment (andrewgarner) [inline features/step_definitions/covering_letter_steps.rb]: Fixed with b3c1075

File: features/step_definitions/record_of_guidance_contents_steps.rb (hunk 3 above)
Comment (andrewgarner) [inline features/step_definitions/record_of_guidance_contents_steps.rb]: Fixed with 8864534

File: features/step_definitions/record_of_guidance_contents_steps.rb (hunk 4 above)
Comment (andrewgarner) [inline features/step_definitions/record_of_guidance_contents_steps.rb]: Fixed with dd35f07

File: app/models/transliterated_appointment_summary.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class TransliteratedAppointmentSummary < SimpleDelegator
+  %i(first_name last_name guider_name
+     address_line_1 address_line_2 address_line_3
+     town county postcode country).each do |method_name|
+    define_method(method_name) do
+      I18n.transliterate(super())
```
Comment (gjvis) [inline app/models/transliterated_appointment_summary.rb]: if we leave off the `()` on `super()` then these will keep functioning if any of these methods ever accept arguments in future

File: features/step_definitions/covering_letter_steps.rb [hunk 8]
```diff
@@ -3,17 +3,13 @@
...
-    attendee_name: output_document.attendee_name,
-    lead: output_document.lead
-  }
+  expected = fixture(:populated_csv).slice(%i(attendee_address_line_1
+                                              attendee_address_line_2
+                                              attendee_address_line_3
+                                              attendee_town
+                                              attendee_county
+                                              attendee_postcode
+                                              attendee_name))
```
Comment (gjvis) [inline features/step_definitions/covering_letter_steps.rb]: :+1: nice use of slice

File: app/models/transliterated_appointment_summary.rb (hunk 7 above)
Comment (andrewgarner) [inline app/models/transliterated_appointment_summary.rb]: Unfortunately that doesn't work, you get the following error `implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.`

Comment (gjvis) [inline app/models/transliterated_appointment_summary.rb]: interesting, didn't know that

**Comment by gjvis** [comment]
Squish and merge? :+1: 

(I know yesterday you mentioned an additional change you wanted to make, but I can't remember what it was)
